### ISSUE TITLE
Feature request: Provide option to disable `fzf-history-widget` line numbers

### ISSUE THREAD

**Issue by nh2** [junegunn/fzf#3007]
- [x] I have read through the manual page (`man fzf`)
- [x] I have the latest version of fzf
- [x] I have searched through the existing issues

## Info

- OS
    - [x] Linux
    - [ ] Mac OS X
    - [ ] Windows
    - [ ] Etc.
- Shell
    - [ ] bash
    - [x] zsh
    - [ ] fish

## Problem / Steps to reproduce

The zsh history invocation [used by fzf](https://github.com/junegunn/fzf/blob/f9d53303bb00225edb6e46ca1851e5cf1b55e623/shell/key-bindings.zsh#L100) `fc -rl 1`, as well as the normal zsh `history` command, output line numbers, which are annoying for most filtering use cases (details below).

Example output of zsh history:

```
 13889  ssh machine1.example.com
 13890  ssh machine2.example.com
 13891  ssh machine3.example.com
 13892  echo hi
```

With this history, searching with Ctrl+R for `ssh machi 1`, you'd expect it to bring up `ssh machine1.example.com`, but it does not, instead bringing up `ssh machine3.example.com`.

This is because the `1` matches the `1` in `13891  ssh machine3.example.com`.

Screenshot:

![image](https://user-images.githubusercontent.com/399535/195901243-29661ea4-36e9-4171-a520-086508600d27.png)

Thus, it would be great to have an option to turn the history line numbers off.

**Comment by nh2** [comment]
This could be achieved (without option) by changing the code like this:

```diff
-  selected=( $(fc -rl 1 | awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, "", cmd); if (!seen[cmd]++) print $0 }' |
+  selected=( $(fc -rln 1 | awk '{ cmd=$0; if (!seen[cmd]++) print $0 }' |
```

where the `-n` flag added to `fc` suppresses the emission of line numbers.

Then we also don't need the code for ignoring line numbers and `*` for deduplication purposes any more, see https://github.com/junegunn/fzf/pull/2140#issuecomment-1279254145 and https://github.com/zsh-users/zsh/blob/b7f280ab597b07bbb22fb741589fe7754fc5cc0c/Src/builtin.c#L1822-L1828.

**Comment by nh2** [comment]
By the way, I'm not quite sure what the purpose of `-n2..,..` is. Why would one limit search scope to a subset of fields, and then also all fields? Is the order relevant here? I cannot conclude the purpose from `fzf --help`'s entry on `-n`.

**Comment by junegunn** [comment]
> searching with Ctrl+R for `ssh machi 1`, you'd expect it to bring up `ssh machine1.example.com`

You just type in `sshmachi1` and get the right one. It's the beauty of the fuzzy matching; better results with less keystrokes.

Spaces are only needed when you later want to add a term that is not in order, e.g. `mach1 ssh`.


> what the purpose of `-n2..,..` is.

`2..` allows prefix matching against the command part, for example`^ssh`, yet `^13891` or `138ssh` are still allowed thanks to `..`.

**Comment by boris-petrov** [comment]
@junegunn thank you for the suggestion but I have another case. Say I have `ssh 192.168.1.234` in my history. I want to type CTRL+R and then 234 and that to choose me that line. However, it would select my 234-th line in the history. Isn't there a way to modify that? I don't really understand why anyone needs to see, let alone search for, history line number?

**Comment by LangLangBart** [comment]
> However, it would select my 234-th line in the history. Isn't there a way to modify that?

You can overwrite  `--nth 2..,..` by setting `FZF_CTRL_R_OPTS`
```sh
# the fist column (event numbers) is excluded from the search
export FZF_CTRL_R_OPTS="--nth 2.."
```

You can also hide the event numbers with `--with-nth 2..`
```sh
export FZF_CTRL_R_OPTS="--with-nth 2.."
```

---

> I don't really understand why anyone needs to see, let alone search for, history line number?

Related comment by the maintainer: https://github.com/junegunn/fzf/pull/3807#issuecomment-2134407188

> junegunn: I prefer to have the numbers. They help me understand the chronological ordering among the matched entries.

**Comment by boris-petrov** [comment]
@LangLangBart thank you very much for the detailed answer! Works like a charm!

**Comment by mbhall88** [comment]
In addition to @LangLangBart's great answer I also found a weird issue on my machine where upon accepting a line in history it also pasted the line number to my terminal, even when line numbers were hidden. What worked for me was

```sh
export FZF_CTRL_R_OPTS="--with-nth=2.. --accept-nth=2.."
```

**Comment by LangLangBart** [comment]
> I also found a weird issue on my machine where upon accepting a line in history it also pasted the line number to my terminal

A reproducible issue was recently reported:

https://github.com/junegunn/fzf/issues/3591#issuecomment-3092322783

It was fixed in version 0.65.0:

https://github.com/junegunn/fzf/pull/4468

If you are already using version 0.65.0 and you source the zsh shell function as usual with `source <(fzf --zsh)` but still experience the issue, posting the results of your debug output might help identify the problem. Don't forget to unset your current `FZF_CTRL_R_OPTS` to not skew the results.

```bash
source <(fzf --zsh)
# Improved trace prompt
PS4='%B%F{0}+%(2e.+.)%(3e.+.)%(4e.+.) %D{%T:%3.}%f %_ [%(?.%F{green}.%F{red})%?%f] %F{cyan}%N%f:%F{yellow}%I%f%b > '
# Enable tracing for the fzf zsh functions of interest
typeset -fT fzf-history-widget __fzf_exec_awk
# Press Ctrl-R, then Enter
```

Copy and paste the last 10–20 lines here. The output should look similar to the example below...

```bash
+ 09:24:32:293 then [0] fzf-history-widget:134 > selected=$'3\ttypeset -fT fzf-history-widget __fzf_exec_awk' 
+ 09:24:32:779  [0] fzf-history-widget:143 > local ret=0
+ 09:24:32:779 if [0] fzf-history-widget:144 > [ -n $'3\ttypeset -fT fzf-history-widget __fzf_exec_awk' ']'
+ 09:24:32:779 then if [0] fzf-history-widget:145 > [[ $(__fzf_exec_awk '{print $1; exit}' <<< "$selected") -regex-match ^[1-9][0-9]*+ 09:24:32:781 then if cond cmdsubst [0] fzf-history-widget:145 > __fzf_exec_awk '{print $1; exit}'
++ 09:24:32:782 if [0] __fzf_exec_awk:276 > [[ -z '' ]]
++ 09:24:32:782 then [0] __fzf_exec_awk:277 > __fzf_awk=awk 
++ 09:24:32:782 then if [0] __fzf_exec_awk:278 > [[ darwin24.2.0 == solaris* ]]
++ 09:24:32:783 then elif [1] __fzf_exec_awk:280 > command -v mawk
++ 09:24:32:784  [0] __fzf_exec_awk:286 > LC_ALL=C awk '{print $1; exit}'
+ 09:24:32:779 then if [0] fzf-history-widget:145 > [[ $(__fzf_exec_awk '{print $1; exit}' <<< "$selected") -regex-match ^[1-9][0-9]* ]]
+ 09:24:32:789 then then [0] fzf-history-widget:146 > zle vi-fetch-history -n 3
+ 09:24:32:789  [0] fzf-history-widget:151 > zle reset-prompt
+ 09:24:32:789  [0] fzf-history-widget:152 > return 0
```
